src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Extensions/SyntaxNodeExtensions.cs

```diff
@@ -385,4 +385,118 @@ public static int GetTrailingWhitespaceLength(this SyntaxNode node, FormattingCo
 
         return node;
     }
+
+    public static bool ExistsOnTarget(this SyntaxNode node, SyntaxNode target)
+    {
+        var matchingNode = target.DescendantNodesAndSelf()
+            // Empty new lines can affect our comparison so we remove them since they're insignificant.
+            .Where(n => n.RemoveEmptyNewLines().ToFullString() == node.RemoveEmptyNewLines().ToFullString())
+            .FirstOrDefault();
+
+        return matchingNode is not null;
+    }
+
+    public static SyntaxNode RemoveEmptyNewLines(this SyntaxNode node)
+    {
+        if (node is MarkupTextLiteralSyntax markupTextLiteral)
+        {
+            var literalTokensWithoutLines = new SyntaxList<SyntaxToken>(markupTextLiteral.LiteralTokens.Where(t => t.Kind != SyntaxKind.NewLine));
+            var updatedLiteral = markupTextLiteral.WithLiteralTokens(literalTokensWithoutLines);
+            return updatedLiteral;
+        }
+
+        return node;
+    }
+
+    public static bool IsCSharpNode(this SyntaxNode node, [NotNullWhen(true)] out CSharpCodeBlockSyntax? csharpCodeBlock)
+    {
+        csharpCodeBlock = null;
+
+        if (node is CSharpCodeBlockSyntax outerCSharpCodeBlock)
+        {
+            var innerCsharpStatement = outerCSharpCodeBlock.ChildNodes().FirstOrDefault(n => n is CSharpStatementSyntax);
+            if (innerCsharpStatement is not null)
+            {
+                return innerCsharpStatement.IsCSharpNode(out csharpCodeBlock);
+            }
+
+            var innerRazorDirective = outerCSharpCodeBlock.ChildNodes().FirstOrDefault(n => n is RazorDirectiveSyntax);
+            if (innerRazorDirective is not null)
+            {
+                return innerRazorDirective.IsCSharpNode(out csharpCodeBlock);
+            }
+
+            var innerCSharpExplicitExpression = outerCSharpCodeBlock.ChildNodes().FirstOrDefault(n => n is CSharpExplicitExpressionSyntax);
+            if (innerCSharpExplicitExpression is not null)
+            {
+                return innerCSharpExplicitExpression.IsCSharpNode(out csharpCodeBlock);
+            }
+
+            var innerCSharpImplicitExpression = outerCSharpCodeBlock.ChildNodes().FirstOrDefault(n => n is CSharpImplicitExpressionSyntax);
+            if (innerCSharpImplicitExpression is not null)
+            {
+                return innerCSharpImplicitExpression.IsCSharpNode(out csharpCodeBlock);
+            }
+        }
+        // @code {
+        //    var foo = "bar";
+        // }
+        else if (node is RazorDirectiveSyntax razorDirective)
```

Comment:
I think we probably want to check the type of directive here. `@attribute [go]` is also a razor directive with a C# body. Or maybe its okay if that matches?